src/server/pbs_db_func.c

```diff
@@ -0,0 +1,588 @@
+#include <signal.h>
+#endif
+
+#define MAX_DB_RETRIES			5
+#define MAX_DB_LOOP_DELAY		10
+#define IPV4_STR_LEN	15
+static int db_oper_failed_times = 0;
+static int last_rc = -1; /* we need to reset db_oper_failed_times for each state change of the db */
+static int conn_db_state = 0;
+extern int pbs_failover_active;
+extern int server_init_type;
+extern int stalone;	/* is program running not as a service ? */
+char conn_db_host[PBS_MAXSERVERNAME+1];	/* db host where connection is made */
+void *svr_db_conn = NULL; /* server's global database connection pointer */
+void *conn = NULL;  /* pointer to work out a valid connection - later assigned to svr_db_conn */
+extern int pbs_decrypt_pwd(char *, int, size_t, char **, const unsigned char *, const unsigned char *);
+extern pid_t go_to_background();
+void *setup_db_connection(char *, int , int );
+static void *get_db_connect_information();
+static int touch_db_stop_file(void);
+static int start_db();
+void stop_db();
+
+/**
+ * @brief
+ *		Checks whether database is down, and if so
+ *		starts up the database in asynchronous mode.
+ *
+ * @return - Failure code
+ * @retval   PBS_DB_DOWN - Error in pbs_start_db
+ * @retval   PBS_DB_STARTING - Database is starting
+ *
+ */
+static int
+start_db()
+{
+	char *failstr = NULL;
+	int rc;
+
+	log_eventf(PBSEVENT_SYSTEM | PBSEVENT_FORCE, PBS_EVENTCLASS_SERVER, LOG_CRIT, msg_daemonname, "Starting PBS dataservice");
+
+	rc = pbs_start_db(conn_db_host, pbs_conf.pbs_data_service_port);
+	if (rc != 0) {
+		pbs_db_get_errmsg(PBS_DB_ERR, &failstr);
+		snprintf(log_buffer, LOG_BUF_SIZE, "%s %s", "Failed to start PBS dataservice.", failstr ? failstr : "");
+		log_eventf(PBSEVENT_SYSTEM | PBSEVENT_ADMIN, PBS_EVENTCLASS_SERVER, LOG_ERR, msg_daemonname, log_buffer);
+		fprintf(stderr, "%s\n", log_buffer);
+		free(failstr);
+		return PBS_DB_DOWN;
+	}
+
+	sleep(1); /* give time for database to atleast establish the ports */
+	return PBS_DB_STARTING;
+}
+
+/**
+ * @brief
+ *		Stop the database if up, and log a message if the database
+ *		failed to stop.
+ *		Try to stop till not successful, with incremental delay.
+ */
+void
+stop_db()
+{
+	char *db_err = NULL;
+	int db_delay = 0;
+	pbs_db_disconnect(svr_db_conn);
+	svr_db_conn = NULL;
+
+	/* check status of db, shutdown if up */
+	db_oper_failed_times = 0;
+	while (1) {
+		if (pbs_status_db(conn_db_host, pbs_conf.pbs_data_service_port) != 0)
+			return; /* dataservice not running, got killed? */
+
+		log_eventf(PBSEVENT_SYSTEM | PBSEVENT_FORCE, PBS_EVENTCLASS_SERVER, LOG_CRIT, msg_daemonname, "Stopping PBS dataservice");
+
+		if (pbs_stop_db(conn_db_host, pbs_conf.pbs_data_service_port) != 0) {
+			pbs_db_get_errmsg(PBS_DB_ERR, &db_err);
+			snprintf(log_buffer, LOG_BUF_SIZE, "%s %s", "Failed to stop PBS dataservice.", db_err ? db_err : "");
+			log_eventf( PBSEVENT_SYSTEM | PBSEVENT_ADMIN, PBS_EVENTCLASS_SERVER, LOG_ERR, msg_daemonname, log_buffer);
```

Comment:
You don't need log_eventf here!

Reply:
I think its good to log to server log since writing just to stderr will not record the cause.